Dev

jsonoh opened the pull request.

Commit:
fixed signout functionality

Commit:
removed links.json dependency

IanSaucy commented:
Can you add a description of what you changed/fixed/added etc? Bullet points are fine

IanSaucy commented on src/screens/LogOut.tsx:
```diff
@@ -10,8 +10,9 @@ function LogOut() {
-            navigate("/login")
         }).catch((error) => {
             // An error happened.
             console.log(error)
```
We should switch to the "firebase" way of doing things and allow firebase to handle the auth persistence. See this doc for info. 

https://firebase.google.com/docs/auth/web/auth-state-persistence

This means removing the need for touching localStorage directly for auth reasons. We will need to update some routing logic based on this.

IanSaucy commented on src/screens/Interests.tsx:
```diff
@@ -42,9 +48,43 @@ function Interests() {
+      try {
+        const res = await fetch(APIUrl + "resource-lists");
+        const json = await res.json();
+        let jsonData = json.data.map((x: any) => {
+          return {
+            category: x.attributes.category,
+            sub_category: x.attributes.sub_category,
+            link: x.attributes.link,
+          };
+        });
```
Since we're not verifying the structure of `x` we should just tell TS what we expect it to be and then you don't have to manually assign things. Ideally, we verify data from the API but....that's a future problem.

IanSaucy commented on src/screens/Interests.tsx:
```diff
@@ -1,13 +1,19 @@
 import { ProgressStepperCompact4 } from "../components/home/Progressbar";
+import { APIUrl } from "./Home";
+
+type interest = {
+  title: string;
+  selected: boolean;
+};
+
```
for style reasons, lets use interfaces. It also means that it should be re-named to `IInterest` since we named anything that's an interface in CamcelCase with a leading uppercase `I` to indicate that it's an interface(not a type of class)

IanSaucy commented on src/screens/Interests.tsx:
```diff
@@ -42,9 +48,43 @@ function Interests() {
+
+        // only extract unique categories
+        const categoryTitles = Array.from(
+          new Set(jsonData.map((x: any) => x.category))
```
If you make jsonData typed you won't need to specify it as an `any` which will be a lot safer

IanSaucy commented on src/screens/Interests.tsx:
```diff
@@ -42,9 +48,43 @@ function Interests() {
   // Store the chips in state
-  const [chips, setChips] = useState({
-    interests: linksJson
+  const [chips, setChips] = useState<{ interests: interest[] }>({
```
Make this an interface instead of an anonymous interface/type

IanSaucy reviewed (changes requested):
Overall good! Just a few TS changes

jsonoh commented:
For the first commit, I changed the code to remove the user data when the sign-out button is pressed. This prevents automatic redirection to the home page upon subsequent logins.
For the second commit, I made two changes to stop reliance on the local links.json file. The first modification was made to home.tsx, where I altered the code to source user interest values from Firebase instead of links.json.
The second part involved a change in interests.tsx. The original code set the interest dataset to a type similar to links.json. I modified it to define the data type and fetch data from the database instead of relying on the local storage.

IanSaucy commented:
bump on my comments @jasonoh1998

jsonoh commented:
Yep fixing them. Think it'll be done by tomorrow.

Commit:
updates to change request

jsonoh commented:
I made some syntax changes, including correctly naming interfaces. I also replaced all instances of 'x: any' in the project with the corresponding interface.

For authentication, the project originally used local storage. I have altered this to utilize getAuth() and browserLocalPersistence for retaining authentication checks. Although this works, I'm not certain if I've implemented it correctly; it might be redundant in certain places.

IanSaucy commented on src/screens/Home.tsx:
```diff
@@ -52,22 +52,15 @@ function Home() {
   const navigate = useNavigate();
   const app = initializeApp(config.firebaseConfig);
```
You should only call this one at app init. Then use https://firebase.google.com/docs/reference/js/v8/firebase.app to get the already instantiated instance.

Commit:
app, db, auth refactored

IanSaucy commented on src/components/AuthRoute.tsx:
```diff
@@ -1,12 +1,12 @@
 
 const AuthRoute: React.FunctionComponent<PropsWithChildren> = (props) => {
     const { children } = props;
-    const auth = getAuth();
```
It's actually safe to call `getAuth()` multiple times unlike `initiliazeapp()` but using the singleton pattern is a-okay too.

https://github.com/firebase/firebase-js-sdk/discussions/5254

IanSaucy commented on src/components/home/Pinned.tsx:
```diff
@@ -3,6 +3,14 @@ import { useEffect, useState } from "react";
+    link: string;
+  };
+};
+
```
This looks like a general interface as it's used in `GetResource.tsx` maybe put it in a shared file and export it for multiple uses?

IanSaucy commented on src/firebase.ts:
```diff
@@ -0,0 +1,8 @@
+
+const app = initializeApp(config.firebaseConfig);
+export const db = getFirestore(app);
+export const auth = getAuth(app);
```
These are the two that are safe to call multiple times if you didn't want to use the singleton pattern like you did here. Singleton pattern is okay though!

IanSaucy commented on src/screens/Home.tsx:
```diff
@@ -79,15 +66,22 @@ function Home() {
 
   // Get the interests from the user profile
-  const fetchdata = async () => {
+  const fetchdata = async (userEmail = "defaultuser@email.com") => {
```
Can you add a comment explaining why you have a default param here? I know it's for anonymous users but it would be good to explain it as such here.

IanSaucy commented on src/screens/Home.tsx:
```diff
@@ -197,7 +202,7 @@ function Home() {
     fetchdata();
-    console.log("fetching data");
+    // console.log("fetching data");
   }, []);
```
If the user has been authenticated then this code shouldn't run at all no? Since it fetches the results for the default user. I know it only runs once, at initial load. But ideally, there should be some conditional logic to indicate what should happen if a user is already authenticated etc.

IanSaucy commented on src/screens/Interests.tsx:
```diff
@@ -1,27 +1,44 @@
-import { config } from '../config/config';
-import { doc, getDoc,setDoc } from "firebase/firestore";
-import linksJson from "../links.json";
+import { db, auth } from "../firebase";
+<<<<<<< HEAD
+=======
+import { getAuth } from "firebase/auth";
+>>>>>>> 0c004b4 (app, db, auth refactored)
+import { doc, setDoc } from "firebase/firestore";
 import { ProgressStepperCompact4 } from "../components/home/Progressbar";
+import { APIUrl } from "./Home";
+
+interface IInterest {
+  title: string;
+  selected: boolean;
+};
+
+interface IResource {
+  attributes: {
+    category: string;
+    sub_category: string;
+    link: string;
+  };
+};
+
```
Merge issue

jsonoh replied:
oops

IanSaucy reviewed (changes requested):
A few more questions/changes

Commit:
types, singleton, merge issue fixed

IanSaucy reviewed (approved):
lgtm
